zeroex/orderwatch/order_watcher_test.go

```diff
@@ -933,6 +933,56 @@ func TestOrderWatcherBatchEmitsAddedEvents(t *testing.T) {
 	require.Len(t, orders, 2)
 }
 
+func TestOrderWatcherCleanup(t *testing.T) {
+	if !serialTestsEnabled {
+		t.Skip("Serial tests (tests which cannot run in parallel) are disabled. You can enable them with the --serial flag")
+	}
+
+	teardownSubTest := setupSubTest(t)
+	defer teardownSubTest(t)
+
+	meshDB, err := meshdb.New("/tmp/leveldb_testing/" + uuid.New().String())
+	require.NoError(t, err)
+
+	ctx, cancelFn := context.WithCancel(context.Background())
+	defer cancelFn()
+	blockWatcher, orderWatcher := setupOrderWatcher(ctx, t, ethRPCClient, meshDB)
+
+	// Create and add two orders to OrderWatcher
+	amount := big.NewInt(10000)
+	signedOrderOne := scenario.CreateZRXForWETHSignedTestOrder(t, ethClient, makerAddress, takerAddress, amount, amount)
+	watchOrder(ctx, t, orderWatcher, blockWatcher, ethClient, signedOrderOne)
+	signedOrderTwo := scenario.CreateZRXForWETHSignedTestOrder(t, ethClient, makerAddress, takerAddress, amount, amount)
+	watchOrder(ctx, t, orderWatcher, blockWatcher, ethClient, signedOrderTwo)
+	signedOrderOneHash, err := signedOrderTwo.ComputeOrderHash()
+	require.NoError(t, err)
+
+	// Set lastUpdate for signedOrderOne to more than defaultLastUpdatedBuffer so that signedOrderOne
+	// does not get re-validated by the cleanup job
+	signedOrderOneDB := &meshdb.Order{}
+	err = meshDB.Orders.FindByID(signedOrderOneHash.Bytes(), signedOrderOneDB)
+	require.NoError(t, err)
+	signedOrderOneDB.LastUpdated = time.Now().UTC().Add(-defaultLastUpdatedBuffer).Add(-1 * time.Minute)
+	err = meshDB.Orders.Update(signedOrderOneDB)
+	require.NoError(t, err)
+
+	// Subscribe to OrderWatcher
+	orderEventsChan := make(chan []*zeroex.OrderEvent, 10)
+	orderWatcher.Subscribe(orderEventsChan)
+
+	// Since no state changes occurred without corresponding events being emitted, we expect
+	// cleanup not to result in any new events
+	err = orderWatcher.Cleanup(ctx, defaultLastUpdatedBuffer)
+	require.NoError(t, err)
+
+	select {
+	case _ = <-orderEventsChan:
+		t.Error("Expected no orderEvents to fire after calling Cleanup()")
```

Comment:
I don't think that this is a sufficient test for the `Cleanup` method. All we are doing here is checking that there weren't any errors and no events were fired. Ideally we should be able to test that the relevant order fields are updated in the database (specifically I'm looking at `LastUpdated`, `FillableTakerAssetAmount`, and `IsRemoved`).

I understand that there are some limitations here due to the fact that we don't have an ERC20 implementation for testing that doesn't implement the proper events. However, I was looking through the `Cleanup` method and I think there is a way around this. Could we choose _not_ to start the `OrderWatcher`, set up the database and Ganache to hold a certain state (e.g. order hasn't been updated in a while and there has been a fill or other on chain event) and then call `Cleanup` without calling `Start`? It looks to me like it is not necessary to call `Start` in order for `Cleanup` to work, but I might be missing something.

Reply:
Unfortunately it's not that easy. The reason for this is that since this PR, blocks only get persisted to the DB once they are processed by the OrderWatcher. When validating orders, we do so at the latest block stored in the DB. In order to run `Cleanup()` at the latest block, that block (and it's events) _must_ have been processed by OrderWatcher. 

We could try to hack around this order of operations by manually adding each block to the DB without first processing it, but this needs to be done in several places and feels too much of a departure over the actual functionality we're trying to test. The right way to make this test case more robust is to use a non-compliant ERC20 token, which I've already created for the V3 snapshot, but that is a pain to also add to a V2 snapshot. I've opened an issue to address this as soon as we drop V2 support: https://github.com/0xProject/0x-mesh/issues/598.

Reply:
Alright I'm fine with this compromise.